go/libraries/doltcore/env/remotes.go

```diff
@@ -114,105 +133,184 @@ func (r *Remote) GetRemoteDBWithoutCaching(ctx context.Context, nbf *types.NomsB
 	return doltdb.LoadDoltDBWithParams(ctx, nbf, r.Url, filesys2.LocalFS, params)
 }
 
+type PushMeta struct {
+	Opts   []*PushOpts
+	Rsr    RepoStateReader
+	Rsw    RepoStateWriter
+	Remote Remote
+	SrcDb  *doltdb.DoltDB
+	DestDb *doltdb.DoltDB
+	TmpDir string
+}
+
 type PushOpts struct {
 	SrcRef      ref.DoltRef
 	DestRef     ref.DoltRef
 	RemoteRef   ref.DoltRef
-	Remote      Remote
 	Mode        ref.UpdateMode
 	SetUpstream bool
+	HasUpstream bool
 }
 
-func NewPushOpts(ctx context.Context, apr *argparser.ArgParseResults, rsr RepoStateReader, ddb *doltdb.DoltDB, force bool, setUpstream bool, pushAutoSetupRemote bool) (*PushOpts, error) {
-	var err error
-	remotes, err := rsr.GetRemotes()
+func NewPushOpts(ctx context.Context, apr *argparser.ArgParseResults, rsr RepoStateReader, ddb *doltdb.DoltDB, force, setUpstream, pushAutoSetupRemote, all bool) ([]*PushOpts, Remote, error) {
+	if apr.NArg() == 0 {
+		return getPushOptsAndRemoteFromNoArg(ctx, rsr, ddb, force, setUpstream, pushAutoSetupRemote, all)
+	}
+
+	rsrBranches, err := rsr.GetBranches()
 	if err != nil {
-		return nil, err
+		return nil, NoRemote, err
```

Comment:
(minor/style) You might consider returning `*Remote` instead of a copy. The pointer allows you to easily return `nil` without having to use special-case instances like `NoRemote`, plus it's slightly more efficient with less memory copying, but that's not a huge deal here. It's not a big deal and really more of style/preference.